modules/consensus/accept.go

```diff
@@ -201,8 +270,38 @@ func (cs *ConsensusSet) AcceptBlock(b types.Block) error {
 	if err != nil {
 		return err
 	}
-	// Broadcast the new block to all peers.
-	peers := cs.gateway.Peers()
-	go cs.gateway.Broadcast("RelayBlock", b, peers)
+	// Broadcast the block to all peers <= v0.5.1 and block header to all peers > v0.5.1.
+	var relayBlockPeers, relayHeaderPeers []modules.Peer
+	for _, p := range cs.gateway.Peers() {
+		if build.VersionCmp(p.Version, "0.5.1") <= 0 {
+			relayBlockPeers = append(relayBlockPeers, p)
+		} else {
+			relayHeaderPeers = append(relayHeaderPeers, p)
+		}
+	}
+	go cs.gateway.Broadcast("RelayBlock", b, relayBlockPeers)
+	go cs.gateway.Broadcast("RelayHeader", b.Header(), relayHeaderPeers)
 	return nil
 }
+
+// managedAcceptHeader will validate a block header and see if it extends the
+// heaviest chain. If a header is valid and would extend the heaviest chain,
+// nil is returned. managedAcceptHeader does not fetch the corresponding block
+// and therefore can't accept the block itself. Just as managedAcceptBlock does
+// not broadcast the block, managedAcceptHeader does not broadcast the block
+// header.
+func (cs *ConsensusSet) managedAcceptHeader(h types.BlockHeader) error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	// Start verification inside of a bolt View tx.
+	err := cs.db.View(func(tx *bolt.Tx) error {
+		// Do not accept a header if the database is inconsistent.
+		if inconsistencyDetected(tx) {
+			return errInconsistentSet
+		}
+
+		// Do some relatively inexpensive checks to validate the header
+		return cs.validateHeader(boltTxWrapper{tx}, h)
+	})
+	return err
+}
```

Comment:
why do you need a managed version of AcceptHeader? The reason you need a managed version of accept block is because the lock has to be demoted partway through the function, which means you can't wrap it in a lock. AcceptHeader on the other hand doesn't push any updates to the modules (nor does it even modify the consensus set), so you don't need to worry about demotion. I don't think this function needs to be here, nor does it need to be managed.